### ISSUE TITLE
Using dynamic username openssh plugin not work

### ISSUE THREAD

**Issue by rlemaitre1970** [rundeck/rundeck#5715]
Issue type: Bug report/Enhancement Request

Hello
Project Configuration : openssh / executor
Version Rundeck 3.2.1-20200113

Replacing the user dynamic variable does not work. After creating an option.username value, the replacement is empty.

The replacement works perfectly with the ssh classic plugin.

Create option username in job :
node username contains ${option.username)
execution result :

[ssh-exec] sshpass -f /tmp/.ssh-exec/ssh-passfile.@lingrh1.ac-lille.fr.tJQjW ssh -p 22 -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -o LogLevel=quiet @lingrh1.ac-lille.fr ls -l

Username Empty

Thank

LEMAITRE Reynald

**My Rundeck detail**

* Rundeck version: 3.2.1-20200113
* install type: deb
* OS Name/version: buster
* DB Type/version: mysql

**Expected Behavior**

**How to Reproduce**
I see the same behavior with this configuration:

1) adding this to project.properties (to get the username via option):

project.ssh-authentication=password
project.ssh-password-storage-path=keys/passwd
project.ssh.user=${option.myusr}

2) With this node source entry:

node name="raspberry" description="Raspberry Pi node" tags="" hostname="10.0.1.28" osArch="armv71" osFamily="unix" osName="Linux" osVersion="4.19.75-v7+" ssh-authentication="password"

3) With this job definition:

<joblist>
  <job>
    <context>
      <options preserveOrder='true'>

        <option name='myusr' value='pi' />

      </options>
    </context>
    <defaultTab>nodes</defaultTab>
    <description></description>
    <dispatch>
      <excludePrecedence>true</excludePrecedence>
      <keepgoing>false</keepgoing>
      <rankOrder>ascending</rankOrder>
      <successOnEmptyNodeFilter>false</successOnEmptyNodeFilter>
      <threadcount>1</threadcount>
    </dispatch>
    <executionEnabled>true</executionEnabled>

    <id>46bedaa7-f9dc-4acb-b942-fc191be3bffe</id>
    <loglevel>INFO</loglevel>
    <name>HelloWorld</name>
    <nodeFilterEditable>false</nodeFilterEditable>
    <nodefilters>
      <filter>name: raspberry</filter>

    </nodefilters>
    <nodesSelectedByDefault>true</nodesSelectedByDefault>
    <scheduleEnabled>true</scheduleEnabled>
    <sequence keepgoing='false' strategy='node-first'>
      <command>

        <exec>echo "hello world"</exec>
      </command>
    </sequence>
    <uuid>46bedaa7-f9dc-4acb-b942-fc191be3bffe</uuid>
  </job>
</joblist>

Using SSH node executor (jsch):

Fail.

Using OpenSSH node executor (with "dry run" to see the command output, if unmark "dry run" the job fails):

[ssh-exec] sshpass -h /tmp/.ssh-exec/ssh-passfile.@10.0.1.28.P60f0 ssh -p 22 -o UserKnownHostsFile=/dev/null -o LogLevel=quiet @10.0.1.28 echo hello world

username empty.

Thank you

**Comment by MegaDrive68k** [comment]
Based on this:

https://github.com/rundeck-plugins/openssh-node-execution#examples

Reproduced using the easiest way to use dynamic user over OpenSSH project:

project.properties:

```
#Project ProjectOPENSSH configuration, generated
#Fri Jan 17 16:48:00 CLST 2020
project.jobs.gui.groupExpandLevel=1
resources.source.2.config.includeServerNode=true
project.ssh-authentication=privateKey
project.plugin.FileCopier.ssh-copier.authentication=password
service.FileCopier.default.provider=ssh-copier
project.execution.history.cleanup.retention.days=60
project.plugin.NodeExecutor.ssh-exec.authentication=password
project.execution.history.cleanup.enabled=false
project.execution.history.cleanup.schedule=0 0 0 1/1 * ? *
project.output.allowUnsanitized=false
project.execution.history.cleanup.batch=500
project.disable.executions=false
resources.source.2.config.writeable=true
project.description=
project.ssh-keypath=/home/m68k/.ssh/id_rsa
service.NodeExecutor.default.provider=ssh-exec
resources.source.2.config.format=resourcexml
resources.source.2.config.file=/home/m68k/Downloads/resources.xml
project.name=ProjectOPENSSH
project.plugin.NodeExecutor.ssh-exec.dry_run=true
project.disable.schedule=false
resources.source.2.config.requireFileExists=true
resources.source.1.type=local
project.label=
resources.source.2.type=file
project.execution.history.cleanup.retention.minimum=50
project.plugin.FileCopier.ssh-copier.dry_run=true
```

resources.xml file:

```
<?xml version="1.0" encoding="UTF-8"?>
<project>
  <node name="hyperion" description="Rundeck server node" tags="" hostname="hyperion" osArch="amd64" osFamily="unix" osName="Linux" osVersion="4.15.0-66-generic" username="m68k"/>
  <node name="Raspberry" description="Remote SSH Node" tags="rb"  hostname="10.0.1.28" osArch="arm64" sFamily="unix" osName="Linux" osVersion="10.12.6" username="${option.myusr}" node-executor="ssh-exec" file-copier="ssh-copier" ssh-authentication="password"  
  ssh-password-storage-path="keys/passwd"/>
</project>
```

Job definition:

```
<joblist>
  <job>
    <context>
      <options preserveOrder='true'>
        <option name='myusr' value='pi' />
      </options>
    </context>
    <defaultTab>nodes</defaultTab>
    <description></description>
    <dispatch>
      <excludePrecedence>true</excludePrecedence>
      <keepgoing>false</keepgoing>
      <rankOrder>ascending</rankOrder>
      <successOnEmptyNodeFilter>false</successOnEmptyNodeFilter>
      <threadcount>1</threadcount>
    </dispatch>
    <executionEnabled>true</executionEnabled>
    <id>6bad7ad2-63b9-41ba-b952-14be5aedd13b</id>
    <loglevel>INFO</loglevel>
    <name>JobTEST</name>
    <nodeFilterEditable>false</nodeFilterEditable>
    <nodefilters>
      <filter>name: Raspberry</filter>
    </nodefilters>
    <nodesSelectedByDefault>true</nodesSelectedByDefault>
    <scheduleEnabled>true</scheduleEnabled>
    <sequence keepgoing='false' strategy='node-first'>
      <command>
        <exec>echo "hi"</exec>
      </command>
    </sequence>
    <uuid>6bad7ad2-63b9-41ba-b952-14be5aedd13b</uuid>
  </job>
</joblist>
```

The option value is empty:

![Screenshot_3](https://user-images.githubusercontent.com/51376003/72642072-4ec33300-394a-11ea-9aa3-3fd9edb56e8e.png)

**Comment by ltamaster** [comment]
@MegaDrive68k 

I think the option must be set as a Secure Remote Authentication

<img width="907" alt="Screenshot 2020-01-17 13 33 02" src="https://user-images.githubusercontent.com/6034968/72647676-f0885700-392d-11ea-9f53-c34dfcabc779.png">

**Comment by MegaDrive68k** [comment]
Thanks @ltamaster 

Using "Secure Remote Authentication".

```

<joblist>
  <job>
    <context>
      <options preserveOrder='true'>
        <option name='myusr' secure='true' value='pi' />
      </options>
    </context>
    <defaultTab>nodes</defaultTab>
    <description></description>
    <dispatch>
      <excludePrecedence>true</excludePrecedence>
      <keepgoing>false</keepgoing>
      <rankOrder>ascending</rankOrder>
      <successOnEmptyNodeFilter>false</successOnEmptyNodeFilter>
      <threadcount>1</threadcount>
    </dispatch>
    <executionEnabled>true</executionEnabled>
    <id>6bad7ad2-63b9-41ba-b952-14be5aedd13b</id>
    <loglevel>INFO</loglevel>
    <name>JobTEST</name>
    <nodeFilterEditable>false</nodeFilterEditable>
    <nodefilters>
      <filter>name: Raspberry</filter>
    </nodefilters>
    <nodesSelectedByDefault>true</nodesSelectedByDefault>
    <scheduleEnabled>true</scheduleEnabled>
    <sequence keepgoing='false' strategy='node-first'>
      <command>
        <exec>echo ${option.myusr}</exec>
      </command>
    </sequence>
    <uuid>6bad7ad2-63b9-41ba-b952-14be5aedd13b</uuid>
  </job>
</joblist>
```

I obtain the same behavior:

![Screenshot_4](https://user-images.githubusercontent.com/51376003/72648015-b41e2080-3958-11ea-8f7d-1eee9836eeb5.png)

maybe I'm forgetting something?

**Comment by rlemaitre1970** [comment]
Hello
no apparently it's a bad behavior of the openssh plugin because in the doc of the plugin it is well indicated the use that we have just configured. In addition with the plugin SSH classic it works perfectly. It detects the variable well but it becomes empty when replaced. If we put another value for the node the value is present in the command line.
Reynald

**Comment by ltamaster** [comment]
@MegaDrive68k
Can you try with `username="option.myusr"` ?

```
<?xml version="1.0" encoding="UTF-8"?>
<project>
  <node name="hyperion" description="Rundeck server node" tags="" hostname="hyperion" osArch="amd64" osFamily="unix" osName="Linux" osVersion="4.15.0-66-generic" username="m68k"/>
  <node name="Raspberry" description="Remote SSH Node" tags="rb"  hostname="10.0.1.28" osArch="arm64" sFamily="unix" osName="Linux" osVersion="10.12.6" username="option.myusr" node-executor="ssh-exec" file-copier="ssh-copier" ssh-authentication="password"  
  ssh-password-storage-path="keys/passwd"/>
</project>
```

**Comment by ltamaster** [comment]
@MegaDrive68k @rlemaitre1970 
Please try this version

https://github.com/rundeck-plugins/openssh-node-execution/releases/tag/2.0.0

it should work with a settings like this one 

```
<?xml version="1.0" encoding="UTF-8"?>
<project>
  <node name="hyperion" description="Rundeck server node" tags="" hostname="hyperion" osArch="amd64" osFamily="unix" osName="Linux" osVersion="4.15.0-66-generic" username="m68k"/>
  <node name="Raspberry" description="Remote SSH Node" tags="rb"  hostname="10.0.1.28" osArch="arm64" sFamily="unix" osName="Linux" osVersion="10.12.6" username="${option.myusr}" node-executor="ssh-exec" file-copier="ssh-copier" ssh-authentication="password"  
  ssh-password-storage-path="keys/passwd"/>
</project>
```

**Comment by MegaDrive68k** [comment]
Hi @rlemaitre1970 

We found a good solution:

1. Download the latest OpenSSH node execution plugin (https://github.com/rundeck-plugins/openssh-node-execution/releases/tag/2.0.0) and put it on `libext` directory.

2. Edit the job and change the option name by "username" (the option now is ${option.username}, I leave my job definition:

```
<joblist>
  <job>
    <context>
      <options preserveOrder='true'>
        <option name='username' value='pi' />
      </options>
    </context>
    <defaultTab>nodes</defaultTab>
    <description></description>
    <dispatch>
      <excludePrecedence>true</excludePrecedence>
      <keepgoing>false</keepgoing>
      <rankOrder>ascending</rankOrder>
      <successOnEmptyNodeFilter>false</successOnEmptyNodeFilter>
      <threadcount>1</threadcount>
    </dispatch>
    <executionEnabled>true</executionEnabled>
    <id>6bad7ad2-63b9-41ba-b952-14be5aedd13b</id>
    <loglevel>INFO</loglevel>
    <name>JobTEST</name>
    <nodeFilterEditable>false</nodeFilterEditable>
    <nodefilters>
      <filter>name: Raspberry</filter>
    </nodefilters>
    <nodesSelectedByDefault>true</nodesSelectedByDefault>
    <scheduleEnabled>true</scheduleEnabled>
    <sequence keepgoing='false' strategy='node-first'>
      <command>
        <exec>echo "hello world"</exec>
      </command>
      <command>
        <exec>ls</exec>
      </command>
    </sequence>
    <uuid>6bad7ad2-63b9-41ba-b952-14be5aedd13b</uuid>
  </job>
</joblist>
```

3. Change the username in node definition accordingly to username option, in my case:

```
<?xml version="1.0" encoding="UTF-8"?>
<project>
  <node name="hyperion" description="Rundeck server node" tags="" hostname="hyperion" osArch="amd64" osFamily="unix" osName="Linux" osVersion="5.3.0-26-generic" username="m68k"/>
  <node name="Raspberry" description="Remote SSH Node" tags=""  hostname="10.0.1.28" osArch="arm64" osFamily="unix" osName="Linux" osVersion="10.12.6" username="${option.username}" node-executor="ssh-exec" file-copier="ssh-copier" ssh-authentication="password" ssh-password-storage-path="keys/passwd"/>
</project>
```

4. Run the job, now works:

![Screenshot_1](https://user-images.githubusercontent.com/51376003/72826108-fa27fc80-3c56-11ea-9111-a9ad8cd375fd.png)




Thanks to @ltamaster 

Regards!

**Comment by rlemaitre1970** [comment]
OK thanks for your answer
Reynald LEMAITRE

**Comment by MegaDrive68k** [comment]
Solved in OpenSSH node executor plugin 2.0.0 out of the box with Rundeck 3.2.2.
